src/Database/Expression/CaseStatementExpression.php

```diff
@@ -0,0 +1,409 @@
+            $this->setTypeMap($typeMap);
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getValue()
+    {
+        return $this->_value;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function value($value, ?string $valueType = null)
+    {
+        if (
+            $value !== null &&
+            !is_scalar($value) &&
+            !(is_object($value) && !($value instanceof Closure))
+        ) {
+            throw new InvalidArgumentException(sprintf(
+                'The `$value` argument must be either `null`, a scalar value, an object, ' .
+                'or an instance of `\%s`, `%s` given.',
+                ExpressionInterface::class,
+                getTypeName($value)
+            ));
+        }
+
+        $this->_value = $value;
+
+        if (
+            $value !== null &&
+            $valueType === null
+        ) {
+            $valueType = $this->_inferType($value);
+        }
+        $this->_valueType = $valueType;
+
+        $this->_isSimpleVariant = true;
+
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getValueType(): ?string
+    {
+        return $this->_valueType;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getWhen(): array
+    {
+        return $this->_when;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function when($when, $type = null)
+    {
+        if ($this->_whenBuffer !== null) {
+            throw new LogicException(
+                'Cannot add new `WHEN` value while an open `when()` buffer is present, ' .
+                'it must first be closed using `then()`.'
+            );
+        }
+
+        if ($when instanceof Closure) {
+            $when = $when(new WhenThenExpression($this->getTypeMap()));
+            if (!($when instanceof WhenThenExpressionInterface)) {
+                throw new LogicException(sprintf(
+                    '`when()` callables must return an instance of `\%s`, `%s` given.',
+                    WhenThenExpressionInterface::class,
+                    getTypeName($when)
+                ));
+            }
+        }
+
+        if ($when instanceof WhenThenExpressionInterface) {
+            $this->_when[] = $when;
+        } else {
+            $this->_whenBuffer = ['when' => $when, 'type' => $type];
+        }
+
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function then($result, ?string $type = null)
+    {
+        if ($this->_whenBuffer === null) {
+            throw new LogicException(
+                'There is no `when()` buffer present, ' .
+                'you must first open one before calling `then()`.'
+            );
+        }
+
+        $whenThen = (new WhenThenExpression($this->getTypeMap()))
+            ->when($this->_whenBuffer['when'], $this->_whenBuffer['type'])
+            ->then($result, $type);
+
+        $this->_whenBuffer = null;
+
+        $this->when($whenThen);
```

Comment:
Couldn't you skip some complexity by appending to `_when` here?

Reply:
Yeah, the method call doesn't seem to be needed, probably some refactoring leftover.